src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsWriteStream.cs

```diff
@@ -0,0 +1,274 @@
+            }
+
+            // Reset the offset since we will start filling up the packet again.
+            WriteBufferOffset = TdsEnums.HEADER_LEN;
+        }
+
+        /// <inheritdoc />
+        public override int Read(byte[] buffer, int offset, int count) => throw new NotSupportedException();
+
+        /// <inheritdoc />
+        public override long Seek(long offset, SeekOrigin origin) => throw new NotSupportedException();
+
+        /// <inheritdoc />
+        public override void SetLength(long value) => throw new NotSupportedException();
+
+        /// <summary>
+        /// Write operation with the byte buffer is not implemented yet.
+        /// </summary>
+        /// <param name="buffer"></param>
+        /// <param name="offset"></param>
+        /// <param name="count"></param>
+        /// <exception cref="NotSupportedException"></exception>
+        public override void Write(byte[] buffer, int offset, int count) 
+            => Write(buffer.AsSpan(offset, count));
+
+        /// <summary>
+        /// An overload for writing a byte to the stream.
+        /// Use this in case the caller is always going to write asynchronously. 
+        /// Otherwise use the WriteByteAsync because it takes the isAsync flag.
+        /// </summary>
+        /// <param name="value">The byte value to be written to the stream.</param>
+        public override void WriteByte(byte value) => WriteByteAsync(value, isAsync: false, ct: CancellationToken.None).AsTask().Wait();
+
+        /// <summary>
+        /// An async mechanism to write a byte to the stream.
+        /// </summary>
+        /// <param name="value">The byte value to be written</param>
+        /// <param name="isAsync">Indicates if the operation is Async</param>
+        /// <param name="ct">Cancellation token to cancel the async operation.</param>
+        /// <returns></returns>
+        public virtual async ValueTask WriteByteAsync(byte value, bool isAsync, CancellationToken ct)
+        {
+            // If we are already at the end of the buffer, flush the buffer, with a softflush.
+            if (_WriteBuffer.Length - WriteBufferOffset == 0)
+            {
+                await FlushAsync(ct, isAsync, FlushMode.SoftFlush).ConfigureAwait(false);
+            }
+            _WriteBuffer[WriteBufferOffset++] = value;
+        }
+
+        /// <inheritdoc/>
+        public override void Write(ReadOnlySpan<byte> buffer)
+        {
+            int len = buffer.Length;
+            // The buffer may not have enough space. Write what we can and then flush the buffer with a soft flush.
+            while (len > 0)
+            {
+                if (len > _WriteBuffer.Length - WriteBufferOffset)
+                {
+                    // Only a part of the length fits in the buffer.
+                    int bytesToWrite = _WriteBuffer.Length - WriteBufferOffset;
+                    buffer.Slice(0, bytesToWrite).CopyTo(_WriteBuffer.AsSpan(WriteBufferOffset));
```

Comment:
How should a 1MB/5MB/10MB varbinary payload be handled, assuming a 64KB buffer size? We'd currently perform 16/80/160 `CopyTo` calls (and possibly another set of similar calls into TCP buffers at the next level down).

Is it possible to refactor `FlushAsync` such that one part of it can flush any `Span`/`Memory` into a packet and `Write`/`WriteAsync` can slice up the incoming payload and call that directly? This would skip the buffer copy unless we know there'd definitely be space for our payload in it.

If a varbinary payload won't ever be that large, this can be ignored - I'm not sure exactly what that type of large UPDATE statement would be transported on the network as.

Reply:
@edwardneal  I agree that this optimization is possible. 
At this point I am going to add a TODO: I need to test a few things with SQL server before I can make this optimization.

For cases where the incoming `Span`/`Memory` is larger than the buffer, If we write the buffer directly to the underlying stream, then first I need to write the header to the underlying stream, and then the incoming `Span`/`Memory` 
This splits the outgoing TDS packet in two parts, header and payload. I want to check how SQL server behaves in those cases. 

Your scenario is valid. A Note: the buffer size never goes above 32768 bytes (protocol spec), and the buffer size is controlled by the connection string. Hence a large varbinary will be split across across multiple TDS packets. 

Your use case is valid with SqlCommand.ExecuteNonQuery with SqlParameters where the parameter values are larger than the buffer size. 

I will do some experiments before making these changes. 

Also based on our experiments, the async interaction with SQL server can benefit a lot compared to the current implementation, esp with large data read. I haven't benchmarked large data writes yet.